### PR TITLE
Feat(RBAC): Add authentication

### LINKED ISSUE #259: Add Permission framework
### Acceptance criteria

What are you trying to do, need? What are the acceptance criteria?

We need to use the [Permission framework](https://backstage.io/docs/permissions/overview) included in Backstage.
This issue is to investigate what can be done with this feature and what are the limitations.

Here are some example scenario:
- How a user can view only their components from the software catalog?
- How only admin can access plugins? ie In the announcement plugin or branding plugin, only admins should be able to update the content
- How to use that data from the Keycloak plugin? Do we have everything needed for authorization?

### PR DESCRIPTION
**Pull request by PatAKnight**
## Description

Adds authentication to the showcase app. In preparation for the RBAC work, we will need some form of authentication to ensure that the permissions are correctly applied and access is granted. This PR adds three authentication providers in the form of GitHub, Guest, and Oauth2Proxy. There is also an additional logic that looks at the `auth.environment` variable within the app-config to determine which sign on providers will be available when running the showcase app.

## Which issue(s) does this PR fix

- Fixes #259 

## PR acceptance criteria

Please make sure that the following steps are complete:

- GitHub Actions are completed and successful
- Unit Tests are updated and passing
- E2E Tests are updated and passing
- Documentation is updated if necessary
- Add a screenshot if the change is UX/UI related

## How to test changes / Special notes to the reviewer
To enable the GitHub and Guest providers the following needs to be added to the configuration file
```yaml
auth:
  environment: development
  providers:
    github:
      development:
        clientId: ${GITHUB_APP_CLIENT_ID}
        clientSecret: ${GITHUB_APP_CLIENT_SECRET}
```

To enable the Oauth2Proxy provider, the following needs to be added to the configuration file
```yaml
auth:
  environment: production
  providers:
    github:
      production:
        clientId: ${GITHUB_APP_CLIENT_ID}
        clientSecret: ${GITHUB_APP_CLIENT_SECRET}
    oauth2Proxy: {}
```
- Notice: that we still need to include the github auth declaration as it is needed for the GitHub plugins

### CODE REVIEW

File: .changeset/chilly-ducks-play.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+'backend': minor
+'app': minor
+---
+
+Adds Authentication to the showcase app. Adds the sign on page to that allows users to know use GitHub, guest, or Keycloak depending on
```
Comment (schultzp2020) [inline .changeset/chilly-ducks-play.md:6]: Needs to be reworded, the second sentence doesn't make sense.

File: packages/backend/src/plugins/auth.ts [hunk 2]
```diff
@@ -37,18 +42,68 @@ export default async function createPlugin(
...
             return ctx.issueToken({
               claims: {
-                sub: userRef, // The user's own identity
-                ent: [userRef], // A list of identities that the user claims ownership through
+                sub: userEntity,
+                ent: ownership,
               },
             });
           },
           // resolver: providers.github.resolvers.usernameMatchingUserEntityName(),
```
Comment (schultzp2020) [inline packages/backend/src/plugins/auth.ts:73]: We should remove this comment.

**Review by schultzp2020 — CHANGES_REQUESTED** [review]
First pass

**Review by schultzp2020 — APPROVED** [review]
LGTM

**Review by christophe-f — APPROVED** [review]

**Pull request merged**